langchain/document_loaders/blockchain_async.py

```diff
@@ -0,0 +1,172 @@
+    Polygon mainnet, and Polygon Mumbai testnet.
+
+    If no BlockchainType is specified, the default is Ethereum mainnet.
+
+    The Loader uses the Alchemy API to interact with the blockchain.
+    ALCHEMY_API_KEY environment variable must be set to use this loader.
+
+    The API returns 100 NFTs per request and can be paginated using the
+    startToken parameter.
+
+    If get_all_tokens is set to True, the loader will get all tokens
+    on the contract.  Note that for contracts with a large number of tokens,
+    this may take a long time (e.g. 10k tokens is 100 requests).
+    Default value is false for this reason.
+
+    Future versions of this loader can:
+        - Support additional Alchemy APIs (e.g. getTransactions, etc.)
+        - Support additional blockain APIs (e.g. Infura, Opensea, etc.)
+    """
+
+    def __init__(
+        self,
+        contract_address: str,
+        blockchainType: BlockchainType = BlockchainType.ETH_MAINNET,
+        api_key: str = "docs-demo",
+        startToken: str = "",
+        get_all_tokens: bool = False,
+        max_execution_time: int = 60,
+    ):
+        self.contract_address = contract_address
+        self.blockchainType = blockchainType.value
+        self.api_key = os.environ.get("ALCHEMY_API_KEY") or api_key
+        self.startToken = startToken
+        self.get_all_tokens = get_all_tokens
+        self.max_execution_time = max_execution_time
+
+        if not self.api_key:
+            raise ValueError("Alchemy API key not provided.")
+
+        if not re.match(r"^0x[a-fA-F0-9]{40}$", self.contract_address):
+            raise ValueError(f"Invalid contract address {self.contract_address}")
+
+    def load(self) -> List[Document]:
+        loop = asyncio.get_event_loop()
+        result = loop.run_until_complete(self._async_load())
+        return result
+
+    async def _async_load(self) -> List[Document]:
+        result = []
+
+        async with aiohttp.ClientSession() as session:
+            if self.get_all_tokens:
+                queue: Queue = Queue()
+                initial_task = self._fetch_data(session, str(0))
+                queue.put_nowait(initial_task)
+
+                while not queue.empty():
+                    task = queue.get_nowait()
+                    items = await task
+
+                    if not items:
+                        break
+
+                    for item in items:
+                        content = str(item)
+                        tokenId = item["id"]["tokenId"]
+                        metadata = {
+                            "source": self.contract_address,
+                            "blockchain": self.blockchainType,
+                            "tokenId": tokenId,
+                        }
+                        result.append(Document(page_content=content, metadata=metadata))
+
+                    next_start_token = self._get_next_tokenId(
+                        result[-1].metadata["tokenId"]
+                    )
```

Comment:
I think this is the blocker of this async idea. Each call has the dependency with the one before... This is also why I removed my previous comment with the AI-generated example... (in case you have received the email notification even though I had removed it...)
From my limited understanding, I think for blockchain async, it's possible when we are fetching with multiple heads of blockchains so that we can process them on parallel, but for each chain, it's a sequential process. But my knowledge of blockchain is limited. The thought above is based on our example logic here. (I also did some exploration online for `Ethereum` and `alchemy api`, but couldn't find a good solution :<)

Reply:
thanks for the feedback.  I'll take another look at this, but unfortunately not until later this week.   The good news is that I think there are 3 approaches:

- the one currently in blockchain.py uses a max_execution_time param that the client can set if 40secs is too long to wait.  I think from a client perspective, this will be ok for most people.
- fixing the async.  currently, it is 'working' in that the test cases are passing, but it is not executing in parallel correctly.  so the request takes too long.
- at some point today, I had the solution 'working' but the requests were failing because they were being sent to the API too quickly and I was hitting rate limits.  Unfortunately, I moved off that approach too quickly and need to remember how I had this setup.  There must be a way to batch or slow-down those requests to prevent the rate limit issue.

Reply:
@skcoirz ok so I think the async is pretty close.  I ran into a few issues with rate limits in the API, then others with representing the correct url value dynamically (we don't know how many async tasks there will be ahead of time). 

I won't have many cycles for a few days to finalize the async work, but i kept a version of it on my fork [here ](https://github.com/saginawj/langchain/blob/get-all-nfts-async/langchain/document_loaders/blockchain_async.py).  In the PR, i updated the sync version so it is working with the timeout, and I believe this model will work for most use cases, though I do want to figure out how to optimize with the async calls.  I also don't want to overcomplicated the doc loader or implement in a way that is not aligned with the other loader, so any feedback here would be appreciated!   I've been through several iterations on the async (e.g. with various strategies for dealing with rate limits on the API), but scaled back in the posted version.

Reply:
yeah sounds good. async is not trivial. We don’t need to rush for it. Thank you for your patience and support in the iterations and discussions here! I really appreciate it!